src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -45,24 +45,45 @@ import scala.collection.mutable.ListBuffer
 
 
 object GroupReadsByUmi {
+  /** A type representing an actual UMI that is a string of DNA sequence. */
+  type Umi = String
+  /** A type to represent a unique ID assigned to a molecule. */
+  type MoleculeId = String
+
   private val SortPositionAttribute        = "__GRs"
   private val LibraryAttribute             = "__GRlb"
   private val InsertEndsAndStrandAttribute = "__Insert"
 
   /** Trait that can be implemented to provide a UMI assignment strategy. */
   private[umi] sealed trait UmiAssigner {
-    def assign(rawUmis: Seq[String]) : Map[String, String]
+    private val counter = new AtomicLong()
+
+    /** Take in a sequence of UMIs and assign each UMI to a unique UMI group ID. */
+    def assign(rawUmis: Seq[Umi]) : Map[Umi, MoleculeId]
+
+    /** Default implementation of a method to retrieve the next ID based on a counter. */
+    protected def nextId: MoleculeId = this.counter.getAndIncrement().toString
+
+    /** Takes in a map of UMI to "sentinel" UMI, and outputs a map of UMI -> Molecule ID. */
+    protected def assignIds(assignments: Map[Umi,Umi]): Map[Umi,MoleculeId] = {
+      val idMap = assignments.values.map(umi => umi -> nextId).toMap
```

Comment:
are the values unique?  It would be nice since then some `MoleculeId`s may not be used.